lib/miq_expression.rb

```diff
@@ -1427,23 +1425,93 @@ def to_arel(exp, tz)
     end
   end
 
-  def extract_where_values(klass, scope)
-    relation = ActiveRecord::Relation.new(klass, klass.arel_table, klass.predicate_builder)
-    relation = relation.instance_eval(&scope)
+  def subquery_for_contains(field, subquery_limiter)
+    # base_class removes the default scopes. They are already in the main query and not needed in this subquery
+    main_model = field.model.base_class
 
-    begin
-      # This is basically ActiveRecord::Relation#to_sql, only using our
-      # custom visitor instance
-
-      connection = klass.connection
-      visitor    = WhereExtractionVisitor.new(connection)
-
-      arel  = relation.arel
-      binds = relation.bound_attributes
-      binds = binds.collect(&:value_for_database)
-      binds.map! { |value| connection.quote(value) }
-      collect = visitor.accept(arel.ast, Arel::Collectors::Bind.new)
-      collect.substitute_binds(binds).join
+    to_ref = field.reflections.first
+    join_keys = to_ref.join_keys
+    primary_attribute = field.model.arel_attribute(join_keys.foreign_key)
+    secondary_attribute = to_ref.klass.arel_attribute(join_keys.key)
+
+    # algorithm:
+    # 1. generate subquery as part of our main table since it has all the concepts we need.
+    # 2. remove main_table from this sub query.
+    # 3. move the secondary_table from the JOIN to the FROM clause
+    # 4. add miq_expression filter to sub query.
+    # 5. if the join on clause has an AND, add the second part back to the query
+    # 6. replace SELECT 1 place holder with secondary_table id
+
+    # 1. SELECT 1 FROM main_table JOIN secondary_table ON main_table.id = secondary_table.main_id AND type = 'type'
+    includes_associations = field.reflections.inject({}) { |i, k| {k.name => i} }
+    query_relation = main_model.joins(includes_associations).select(1).distinct
+    query = query_relation.arel
+
+    # 2. SELECT 1 FROM main_table [ ]
```

Comment:
Should this comment be:

```ruby
# 2. SELECT 1 FROM [ ]  (drop main query)
```

Or, instead is the algorithm supposed to be:

```ruby
    # 2. drop joins from sub query
    # 3. replace the FROM with the secondary_table from the JOIN
```

Looks like we are doing the latter, meaning the comment above makes sense as is, but not sure.

Reply:
agreed, this is confusing. I moved stuff around in attempt to address this